cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java

```diff
@@ -185,97 +192,88 @@ public boolean execute(String action, JSONArray args, CallbackContext callbackCo
-                try {
-                  // Tunnel instance actions
-                  if (Action.START.is(action)) {
-                    // Set instance variables in case we need to start the VPN service from
-                    // onActivityResult
-                    startRequestTunnelId = tunnelId;
-                    startRequestConfig = args.getJSONObject(1);
-                    prepareAndStartVpn();
-                  } else if (Action.STOP.is(action)) {
-                    stopVpnTunnel(tunnelId);
-                  } else if (Action.IS_REACHABLE.is(action)) {
-                    boolean isReachable =
-                        ShadowsocksConnectivity.isServerReachable(
-                            args.getString(1), args.getInt(2));
-                    PluginResult result = new PluginResult(PluginResult.Status.OK, isReachable);
-                    sendPluginResult(tunnelId, action, result, false);
-                  } else if (Action.IS_RUNNING.is(action)) {
-                    PluginResult result =
-                        new PluginResult(PluginResult.Status.OK, isTunnelActive(tunnelId));
-                    sendPluginResult(tunnelId, action, result, false);
-
-                    // Static actions
-                  } else if (Action.INIT_ERROR_REPORTING.is(action)) {
-                    final String apiKey = args.getString(0);
-                    OutlineLogger.initializeErrorReporting(getBaseContext(), apiKey);
-                    callback.success();
-                  } else if (Action.REPORT_EVENTS.is(action)) {
-                    final String uuid = args.getString(0);
-                    OutlineLogger.sendLogs(uuid);
-                    callback.success();
-                  } else {
-                    LOG.severe(
-                        String.format(Locale.ROOT, "Unexpected asynchronous action %s", action));
-                    callback.error(ErrorCode.UNEXPECTED.value);
-                  }
-                } catch (Exception e) {
-                  LOG.log(Level.SEVERE, "Unexpected error while executing action.", e);
-                  if (isTunnelInstanceAction(action)) {
-                    PluginResult pluginResult =
-                        new PluginResult(PluginResult.Status.ERROR, ErrorCode.UNEXPECTED.value);
-                    sendPluginResult(tunnelId, action, pluginResult, false);
-                  } else {
-                    callback.error(ErrorCode.UNEXPECTED.value);
-                  }
-                }
-              }
-            });
-  }
-
-  private void prepareAndStartVpn() {
-    if (prepareVpnService()) {
-      startVpnTunnel();
-    }
+  protected void executeAsync(
+      final String action, final JSONArray args, final CallbackContext callback) {
+    cordova.getThreadPool().execute(() -> {
+      try {
+        // Tunnel instance actions: tunnel ID is always the first argument.
+        if (Action.START.is(action)) {
+          if (!prepareVpnService()) {
```

Comment:
Could you move this check to before spawning a new thread?  Spawning a new thread for blocking calls, and then converting this intrinsically asynchronous operation into blocking so it can run on that thread, seems complicated.

Reply:
I'm suggesting not to wait, and to handle this as an async operation directly, instead of converting it to a blocking operation that runs on a fresh thread.  This would replace `vpnPreparedSignal` (fancy multithreading stuff) with a `CallbackContext` (straightforward async callback).  I think blocking calls and thread-per-call make sense for communication with the VPN process, but we don't need to use the same pattern for the permission intent.

If you want the 30-second-delay behavior, that might be possible with Looper.getMainLooper() and Handler.postDelayed().